waiter/src/waiter/correlation_id.clj

```diff
@@ -41,24 +41,28 @@
 (def default-correlation-id "UNKNOWN")
 (def ^:dynamic dynamic-correlation-id default-correlation-id)
 
+(defn get-correlation-id
+  "Retrieve the value of the current correlation-id."
+  []
+  dynamic-correlation-id)
+
 (defmacro with-correlation-id
   "Executes the body with the specified value of correlation-id."
   [correlation-id & body]
   `(binding [dynamic-correlation-id ~correlation-id]
-     (let [start-thread-id# (.getId (Thread/currentThread))
+     (let [start-thread# (Thread/currentThread)
            result# (do ~@body)
-           end-thread-id# (.getId (Thread/currentThread))]
-       (when (not= start-thread-id# end-thread-id#)
+           end-thread# (Thread/currentThread)]
+       (when (not= (.getId start-thread#) (.getId end-thread#))
          (log/warn "with-correlation-id binding executed on different threads"
-                   {:correlation-id ~correlation-id
-                    :thread-ids {:end end-thread-id# :start start-thread-id#}}))
+                   {:end {:correlation-id (get-correlation-id)
+                          :thread-id (.getId end-thread#)
+                          :thread-name (.getName end-thread#)}
+                    :start {:correlation-id ~correlation-id
```

Comment:
I'd prefer it if we created a binding `(let [correlation-id# ~correlation-id] ...)` at the top level (outside the `binding` block) so that we don't reuse the `correlation-id` argument (which might be an expression rather than a value).

Reply:
Done.